File: DOMTester/xml-generate.py
#!/usr/bin/env python3.1

'''
Python script to generate a schema-compliant XML file from another file
containing a set of points. Input file must have one point per line, defined
as latitude and longitude values separated by a single space. There can be
no blank lines in the file. Sample input file:

39.485189 -87.295395
39.485582 -87.300839
39.485723 -87.303253

Input file must be called points.txt; resulting generated output is points.xml.
'''

with open("points.xml", 'w') as out:
    out.write('<?xml version="1.0" encoding="ISO-8859-1"?><data><trails><trail id="0" name="Heritage"><points>\n')
    with open("points.txt") as f:
        i = 0
        lines = f.readlines()
        for line in lines:
            parts = line.split(" ")
            latitude = parts[0]
            longitude = parts[1].replace("\n", "")
            out.write('<point id="{0}"><title>Point {0}</title><description>Point {0}</description><category id="0">Category</category><condition>Open</condition><latitude>{1}</latitude><longitude>{2}</longitude>'.format(i, latitude, longitude))
            if not i == (len(lines) - 1):
                out.write('<connections><connection>{0}</connection></connections>'.format(i+1))
            out.write('</point>\n')
            i = i + 1
    out.write('<misc></misc></data>\n')

File: DOMTester/xml-generate.rb
File.open("points.xml", 'w') do |out_file|
  write_header(out_file)

  in_file = File.new("points.txt", 'r')
  count = 0

  while (line = in_file.gets)
    parts = line.split(" ").map { |part| part.chomp }
    lat = parts[0]
    long = parts[1]
    write_point(lat, long, count, out_file)
    write_connection(out_file, count - 1) if count > 0
    close_point_tag(out_file)
    count += 1
  end

  close_tags(out_file)
end

def write_point(lat, long, id, file)
    file.write("<point id='#{id}'><title>Point #{id}</title><description>Point #{id}</description><category id="0">Category</category><condition>Open</condition><latitude>#{lat}</latitude><longitude>#{long}</longitude>")
end

def write_connection(file, id)
  file.write("<connections><connection>#{id}</connection></connections>")
end

def write_header(file)
  file.write('<?xml version="1.0" encoding="ISO-8859-1"?><data><trails><trail id="0" name="Heritage"><points>\n')
end

def close_tags(file)
  file.write("</points></trail></trails><misc></misc></data>\n")
end

def close_point_tag(file)
  file.write("</point>\n")
end
